homeassistant/components/binary_sensor/trend.py

```diff
@@ -106,32 +135,51 @@ def device_class(self):
         return self._device_class
 
     @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        return {
+            ATTR_ENTITY_ID: self._entity_id,
+            ATTR_FRIENDLY_NAME: self._name,
+            ATTR_INVERT: self._invert,
+            ATTR_GRADIENT: self._gradient,
+            ATTR_MIN_GRADIENT: self._min_gradient,
+            ATTR_SAMPLE_DURATION: self._sample_duration,
+            ATTR_SAMPLE_COUNT: len(self.samples),
+        }
+
+    @property
     def should_poll(self):
         """No polling needed."""
         return False
 
     @asyncio.coroutine
     def async_update(self):
         """Get the latest data and update the states."""
-        if self.from_state is None or self.to_state is None:
-            return
-        if (self.from_state.state == STATE_UNKNOWN or
-                self.to_state.state == STATE_UNKNOWN):
+        import math
+        import numpy as np
+
+        # Remove outdated samples
+        if self._sample_duration == 0:
+            self.samples = self.samples[-2:]
+        else:
+            now = datetime.now(timezone.utc).timestamp()
+            cutoff = now - self._sample_duration
+            self.samples = [(t, v) for t, v in self.samples if t >= cutoff]
+
+        if len(self.samples) < 2:
             return
-        try:
-            if self._attribute:
-                from_value = float(
-                    self.from_state.attributes.get(self._attribute))
-                to_value = float(
-                    self.to_state.attributes.get(self._attribute))
-            else:
-                from_value = float(self.from_state.state)
-                to_value = float(self.to_state.state)
-
-            self._state = to_value > from_value
-            if self._invert:
-                self._state = not self._state
-
-        except (ValueError, TypeError) as ex:
-            self._state = None
-            _LOGGER.error(ex)
+
+        # Calculate gradient of linear trend
```

Comment:
Move that out of the update in a new function they run in a thread with `self.hass.add_job`